### ISSUE TITLE
Type Instability in Shifts for BitIntegers less than 128 Bit

### ISSUE THREAD

**Issue by PaulDepping** [rfourquet/BitIntegers.jl#27]
I've been getting performance problems with Shifts regarding BitIntegers less than 128 Bit large. Expected behaviour was no allocations and Type Stability, and after some searching, I think I've found the function where it comes from:
```julia
# everything working as expected
julia> @code_warntype UInt136(1) >> Unsigned(3)
MethodInstance for >>(::UInt136, ::UInt64)
  from >>(x::BitIntegers.AbstractBitUnsigned, y::Union{UInt128, UInt16, UInt32, UInt64, UInt8, BitIntegers.AbstractBitUnsigned}) in BitIntegers at /home/pd/.julia/packages/BitIntegers/ukb1R/src/BitIntegers.jl:336
Arguments
  #self#::Core.Const(>>)
  x::UInt136
  y::UInt64
Body::UInt136
1 ─ %1 = BitIntegers.shift_call(BitIntegers.lshr_int, x, y)::UInt136
└──      return %1
```
However, when calling this with different types:
```julia
julia> @code_warntype UInt24(1) >> Unsigned(3)
MethodInstance for >>(::UInt24, ::UInt64)
  from >>(x::BitIntegers.AbstractBitUnsigned, y::Union{UInt128, UInt16, UInt32, UInt64, UInt8, BitIntegers.AbstractBitUnsigned}) in BitIntegers at /home/pd/.julia/packages/BitIntegers/ukb1R/src/BitIntegers.jl:336
Arguments
  #self#::Core.Const(>>)
  x::UInt24
  y::UInt64
Body::Any
1 ─ %1 = BitIntegers.shift_call(BitIntegers.lshr_int, x, y)::Any
└──      return %1
```

This seems to lead to much worse performance when shifting something like an UInt24, compared to the larger BitIntegers.
```julia
julia> @benchmark $(UInt136(1)) >> $(Unsigned(3))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  4.114 ns … 61.751 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     4.157 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   4.289 ns ±  1.847 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▅██▄     ▂▁       ▁       ▁▁       ▁▂▂                    ▂
  █████▆▄▁▄███▆▁▁▁▁▅▇█▇▃▁▁▁▄▇██▇▃▁▁▁▁▅███▆▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▇ █
  4.11 ns      Histogram: log(frequency) by time     4.98 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark $(UInt24(1)) >> $(Unsigned(3))
BenchmarkTools.Trial: 10000 samples with 979 evaluations.
 Range (min … max):  63.586 ns …  1.432 μs  ┊ GC (min … max): 0.00% … 94.89%
 Time  (median):     66.771 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   71.196 ns ± 39.332 ns  ┊ GC (mean ± σ):  1.62% ±  2.96%

  ▄▆█▄▁▁▁ ▁▂▄▃▁                                               ▁
  █████████████▇██▇▆▆▇█▇█▇▆▇▆▇▆▆▆▆▆▆▆▆▅▄▅▄▅▁▄▃▄▁▁▄▄▃▄▄▄▃▁▃▃▃▅ █
  63.6 ns      Histogram: log(frequency) by time       141 ns <

 Memory estimate: 48 bytes, allocs estimate: 3.
```

I've been looking at the generated function `shift_call` to figure out exactly where this type instability is coming from, but I thought I'd share beforehand as well.

Edit: Here is my Julia Version.
```julia
julia> versioninfo()
Julia Version 1.7.2
Commit bf53498635* (2022-02-06 15:21 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-1035G1 CPU @ 1.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, icelake-client)

```

**Comment by PaulDepping** [comment]
It seems to me like this is because of passing a `Core.IntrinsicFunction` to another function. Unlike regular functions, it seems like IntrinsicFunctions all have the same type and thus there is no specialization done for it. Why this does not happen with sizes > 128 bit is beyond me though...

```julia
julia> function pd_shl_int(a, b)
           Core.Intrinsics.shl_int(a, b)
       end
pd_shl_int (generic function with 1 method)

julia> callfunc(a::Function, b, c) = a(b, c)
callfunc (generic function with 1 method)

julia> @code_warntype callfunc(pd_shl_int, 1, 0x2)
Variables
  #self#::Core.Const(callfunc)
  a::Core.Const(pd_shl_int)
  b::Int64
  c::UInt8

Body::Int64
1 ─ %1 = (a)(b, c)::Int64
└──      return %1

julia> @code_warntype callfunc(Core.Intrinsics.shl_int, 1, 0x2)
Variables
  #self#::Core.Const(callfunc)
  a::Core.IntrinsicFunction
  b::Int64
  c::UInt8

Body::Any
1 ─ %1 = (a)(b, c)::Any
└──      return %1
```

**Comment by jlapeyre** [comment]
I found that the same bug. I see it when the number of bits to shift by is not a literal.
> Unlike regular functions, it seems like IntrinsicFunctions all have the same type

I also suspected this but did not verify it.
For me, removing this line
https://github.com/rfourquet/BitIntegers.jl/blob/e669f38050f8feadb1de74bd2d51beaeb91f91ba/src/BitIntegers.jl#L323
seems to fix the bug.

Note that this is a different fix from the one in #27
